packages/contracts/src/contracts/current/forwarder/MixinBuyExactAssets.sol

```diff
@@ -0,0 +1,155 @@
+        require(
+            isAcceptableThreshold(
+                remainingTakerAssetAmount,
+                totalFillResults.takerAssetFilledAmount),
+            UNACCEPTABLE_THRESHOLD
+        );
+        withdrawPayAndDeductFee(remainingTakerAssetAmount, totalFillResults.takerAssetFilledAmount, feeProportion, feeRecipient);
+        return totalFillResults;
+    }
+
+    /// @dev Buys an exact amount of an ERC20 token using WETH.
+    /// @param orders Orders to fill. The maker asset is the ERC20 token to buy. The taker asset is WETH.
+    /// @param signatures Proof that the orders were created by their respective makers.
+    /// @param feeOrders to fill. The maker asset is ZRX and the taker asset is WETH.
+    /// @param feeSignatures Proof that the feeOrders were created by their respective makers.
+    /// @param assetAmount Amount of the ERC20 token to buy.
+    /// @return totalFillResults Aggregated fill results of buying the ERC20 and ZRX tokens.
+    function buyExactERC20TokensInternal(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 assetAmount
+    )
+        private
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        address makerTokenAddress = readAddress(orders[0].makerAssetData, 0); 
+        // We can short cut here for effeciency and use buyFeeTokensInternal if maker asset token is ZRX
+        // this buys us exactly that amount taking into account the fees. This saves gas and calculates the rate correctly
+        Exchange.FillResults memory requestedTokensResults;
+        if (makerTokenAddress == address(ZRX_TOKEN)) {
+            requestedTokensResults = buyFeeTokensInternal(orders, signatures, assetAmount);
+            // When buying ZRX we round up which can result in a small margin excess
+            require(requestedTokensResults.makerAssetFilledAmount >= assetAmount, UNACCEPTABLE_THRESHOLD);
+        } else {
+            Exchange.FillResults memory calculatedMarketBuyResults = calculateMarketBuyFillResults(orders, assetAmount);
+            if (calculatedMarketBuyResults.takerFeePaid > 0) {
+                // Fees are required for these orders. Buy enough ZRX to cover the future market buy
+                Exchange.FillResults memory feeTokensResult = buyFeeTokensInternal(
+                    feeOrders, feeSignatures, calculatedMarketBuyResults.takerFeePaid);
+                totalFillResults.takerAssetFilledAmount = feeTokensResult.takerAssetFilledAmount;
+                totalFillResults.takerFeePaid = feeTokensResult.takerFeePaid;
+            }
+            // Make our market buy of the requested tokens with the remaining balance
+            requestedTokensResults = EXCHANGE.marketBuyOrders(orders, assetAmount, signatures);
+            require(requestedTokensResults.makerAssetFilledAmount == assetAmount, UNACCEPTABLE_THRESHOLD);
```

Comment:
That check is only because we can overbuy ZRX due to rounding. That is, we always want to buy more rather than buy less. If we buy less we cannot fully satisfy the order, so we overbuy slightly.